Add docker plugin upgrade

cpuguy83 opened the pull request:
This allows a plugin to be upgraded without requiring to
uninstall/reinstall a plugin.
Since plugin resources (e.g. volumes) are tied to a plugin ID, this is
important to ensure resources aren't lost.

The plugin must be disabled while upgrading (errors out if enabled).
This does not add any convenience flags for automatically
disabling/re-enabling the plugin during before/after upgrade.

Since an upgrade may change requested permissions, the user is required
to accept permissions just like `docker plugin install`.

Fixes #28913

ping @anusha-ragunathan @tiborvass 
Not setting any milestones at the moment, but it seems like this is needed for production use of plugins.

tiborvass commented:
I thought about this too.

Images with different names (including tag) are different images. There is no notion of "version" or any other upgrade guarantee. So what we're actually talking about is: can we change the driver for existing objects such as volumes?

Anyway, I don't have a better idea so I guess +1 for this design with `plugin upgrade`. Although maybe it could be `docker plugin install --replace $OLD $NEW`.

srust commented:
I agree, I think the `replace` functionality is equally important, for different tags of the same driver. Without replace, a user trying another image (driver:test-v2) no longer has access to existing volumes? Although a docker restart might fix it, restart shouldn't be required.. 

They're really the same volumes, regardless of tag, for the same base driver image, in my opinion.

cpuguy83 commented:
I'm recalling past arguments over `docker run --replace` ;)

srust commented:
I wasn't around for that discussion but sounds exciting. :-)

TBH, I'm not sure why a plugin doesn't specify its "name" that it provides in the config manifest. Then whatever the image name/tag is doesn't matter, its the name of the plugin that it provides. Only one plugin could be loaded at a given time that provides that "name". Internally in docker, the installed plugin is tied to this "name" (not the plugin ID). You could even do the disable, remove, add, enable workflow (without upgrade), and not lose access to your volumes.

But maybe that's just me! :-) There must be a requirement I'm missing for using the plugin ID the way it is now? I'm sure other plugin types don't have this problem, but volumes being persistent make the plugin ID a bit challenging in this case...

cpuguy83 commented:
So, I'm not sure where to go from here... Also, seems like we need something like this for 1.13... but I don't want to paint us into a corner on UX here.

srust commented:
@cpuguy83 First, let me say thanks for taking a stab at this. I understand clearly that code talks, and words are cheap. So I definitely appreciate you working to address this.

That said, the more I think about this, the more I think that the plugin developer should be allowed/required to set a `pluginName`. The name would be used to map volumes to the plugin internally in docker, in place of the `pluginID`, and would also be displayed to the user.

The plugin name would be specified in the `config.json` manifest by the plugin developer, such as:

````
{
    "pluginName": "volume-plugin",
}
````

Volumes would be associated with the `pluginName` internally.

A `docker plugin ls` shows the name and the image:

````
NAME              IMAGE            TAG             DESCRIPTION     ENABLED
volume-plugin     repo/myvolumes   latest          Volume Plugin   false
````

I don't think this would be a huge change in the code, but has the following advantages:

* The plugin developer chooses the name that binds volumes to the plugin: not an arbitrary pluginID.
* The plugin name is de-coupled from an image tag. Why bind volumes to arbitrary or versioned tags. **The plugin developer knows which tags are compatible with each other** and sets the plugin name appropriately. Volumes are more than likely usable from any plugin version with the same name -- your data is not tied to a particular version.
* **Multiple plugins from the same image are supported**. The plugin developer can use the same image but provide it a different name. For example, a volume plugin using the same image could provide `volumes-option1`, `volumes-option2`, etc. I know another issue filed requested this capability for a particular volume plugin (#27299)
* **Backwards compatibility for users**: everyone is used to specifying a plugin by name, such as `blockbridge`, `convoy`, `flocker`, `rexray`, `sshfs`, `nfs`, `cifs`, etc. The plugin developer can set this name to be the same, and existing scripts still work, user expectations are the same.
* **Upgrade path is straightforward**. Can simply remove the old plugin, and install/pull a new one. The plugin name stays the same, the plugin image tag/version can change without affecting your data.
* `docker plugin upgrade` is just gravy. An easy way to pull a new image in place: it becomes a nice helper.

Reasons against this change?

* The plugin code would have to change
* It is late in 1.13
* others?

As a developer of a plugin that's been containerized since august 2015, this list of advantages makes a lot of sense to me.

Thanks for your consideration and feedback.

LK4D4 commented:
@cpuguy83 1.13 was released
Should we proceed with this?

cpuguy83 commented:
🤷‍♂️ on what's right here, but likely something like this is still needed.

cpuguy83 commented:
Let's close for now. We can still discuss, but I doubt this will be the final design.

anusha-ragunathan commented:
We can repurpose this PR to accomodate a clean design. Reopening it.

anusha-ragunathan commented on cli/command/plugin/upgrade.go:
```diff
@@ -0,0 +1,41 @@
+func newUpgradeCommand(dockerCli *command.DockerCli) *cobra.Command {
+	var options pluginOptions
+	cmd := &cobra.Command{
+		Use:   "upgrade [OPTIONS] PLUGIN",
```
Upgrade have options similar to install. - disable, --alias string, --disable-content-trust, [KEY=VALUE...]. Better to have a common function for both.

anusha-ragunathan commented on integration-cli/docker_cli_plugins_test.go:
```diff
@@ -251,3 +251,30 @@ func (s *DockerSuite) TestPluginInspect(c *check.C) {
+	dockerCmd(c, "plugin", "disable", pNameWithTag)
+	dockerCmd(c, "plugin", "upgrade", "--grant-all-permissions", pNameWithTag)
+
+	stat, err = os.Stat(filepath.Join(dockerBasePath, "plugins", id, "rootfs"))
```
A more effective test would be to create volume before and check for existence after upgrade.

anusha-ragunathan commented on plugin/backend_linux.go:
```diff
@@ -312,6 +312,70 @@ func (pm *Manager) Set(name string, args []string) error {
+	pluginDir := filepath.Join(pm.libRoot, p.PluginObj.ID)
+	for _, f := range []string{"rootfs", "config.json", "plugin-settings.json"} {
+		orig := filepath.Join(pluginDir, f)
+		backup := filepath.Join(pluginDir, f+"-old")
```
`backup := orig+"-old"` ?

anusha-ragunathan commented on plugin/backend_linux.go:
```diff
@@ -312,6 +312,70 @@ func (pm *Manager) Set(name string, args []string) error {
+		defer func() {
+			if err != nil {
+				if renameErr := os.Rename(backup, orig); renameErr != nil {
+					logrus.WithField("plugin", p.Name).WithError(err).Errorf("error restoring plugin from backup for file %q", orig)
```
`rootfs` is a dir. So better to omit "file" from error msg.

cpuguy83 commented:
@srust 

In this case I don't think we want the developer to choose the name, as there can be no guarantee uniqueness.

The plugin name is at a technical level decoupled from the image reference, see `--alias` flag, although we need to do more here.  Specifically we need to store the image reference in addition to name/alias (when applicable).

Multiple plugins from the same image are supported (probably after you made the last comment) via `--alias`.
Users can choose back-compat by specifying `--alias`, though maybe it's good to add this to your own plugin docs with the correct alias to use. This can get weird really quickly though if the old plugin is still running.

After speaking with @anusha-ragunathan, we decided to go forward with this design.
This is also extensible in the future to do something like `docker plugin upgrade --image foo awesome/plugin`.

Thanks 🙇 👼

cpuguy83 committed:
Add docker plugin upgrade

This allows a plugin to be upgraded without requiring to
uninstall/reinstall a plugin.
Since plugin resources (e.g. volumes) are tied to a plugin ID, this is
important to ensure resources aren't lost.

The plugin must be disabled while upgrading (errors out if enabled).
This does not add any convenience flags for automatically
disabling/re-enabling the plugin during before/after upgrade.

Since an upgrade may change requested permissions, the user is required
to accept permissions just like `docker plugin install`.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Ok, I have pushed an updated implementation.
Taking into account content-trust which must be handled on the client side, I have had to allow upgrading to (currently) arbitrary plugin images since the API needs to support the client passing through a remote to pull, it's really only a matter of adding the option to the CLI (which I did here)... but it's also supported to not specify anything and it just reads the image reference from the plugin inspect output.

We may want to ask the user to verify that the passed in image is what they really want... maybe contingent only on if it is a whole separate repo vs an exact match to repo+tag. Thoughts?
Maybe check interfaces implemented by the plugin as well?


I can beef up the tests once we are done with design review.

anusha-ragunathan commented:
Few comments:

- It might be valuable to display `ImageReference` as part of plugin ls.

- I agree that if the user specifies a repo different from the `ImageReference`, we should throw a warning. Also, have a corresponding flag to use in CLI.

- A bit concerned about deprecating `--alias`  (i.e hiding it) and replacing with `--name`

- Throw error about "enabled" plugin *before* calculating and prompting for permissions.

- Not sure what's causing it, but I noticed that the timestamp of the rootfs for `cpuguy83/docker...` was off during an upgrade. Didnt test other plugins.

cpuguy83 commented:
> It might be valuable to display ImageReference as part of plugin ls

Agreed, but maybe we can do this separately?

> I agree that if the user specifies a repo different from the ImageReference, we should throw a warning. Also, have a corresponding flag to use in CLI.

👍 

> A bit concerned about deprecating --alias (i.e hiding it) and replacing with --name

The reason for this change is mostly to be less confusing, but for sure we don't need to do it.

> Throw error about "enabled" plugin before calculating and prompting for permissions.

👍 

> Not sure what's causing it, but I noticed that the timestamp of the rootfs for cpuguy83/docker... was off during an upgrade. Didnt test other plugins.

Can you explain what you mean here?

anusha-ragunathan commented:
Timestamp is a non-issue.

Design LGTM.

cpuguy83 commented:
This has been updated with the latest requests.
Need to work out terminology/flag names.

I did not add "image" to the default output just because it makes it super wide, but it's available in the formatter, and is part of the `raw` format.

anusha-ragunathan commented:
@cpuguy83 : `docker_cli_plugins_test.go` doesnt have the original upgrade test. Can you please add it back along with the suggested volume upgrade bits? i.e create a volume using a volume plugin and verify that the volumes are preserved before and after the upgrade.

vieux commented:
what's the point of keeping the old rootfs around ?

also, because of ^, it's not possible to upgrade twice:

```
rename /var/lib/docker/plugins/ff569820f77c0241f7125a6e9893d64c3ba317d45912e53d8cb1015d2e3f5db3/rootfs /var/lib/docker/plugins/ff569820f77c0241f7125a6e9893d64c3ba317d45912e53d8cb1015d2e3f5db3/rootfs-old: directory not empty
```

cpuguy83 commented:
@vieux We should be removing the old rootfs at the end. I've done tons and tons of upgrades against the same plugin with no issue.

Currently trying to work out an issue with the request to have propagated mount be external from rootfs.

cpuguy83 committed:
Make propagated mount persist outside rootfs

This persists the "propagated mount" for plugins outside the main
rootfs. This enables `docker plugin upgrade` to not remove potentially
important data during upgrade rather than forcing plugin authors to hard
code a host path to persist data to.

Also migrates old plugins that have a propagated mount which is in the
rootfs on daemon startup.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
PTAL

vieux commented:
@cpuguy83 test failure

vieux commented:
thanks to @tiborvass's help, I was able to update my plugin to support upgrades. (it basically drops a state.json on the host and we the plugin is upgraded it builds its internal state from the file)

@cpuguy83 we also had issues with propagated mounts during testing, looks like it happens only when propagated mounts and bind mounts overlap, so I avoided that to make it work without leaks.

(btw the issue I mentioned above was probably due to mount leaks, the remove of the -old rootfs was failing)

Pushed the plugin to `vieux/sshfs:next` and the code is here: https://github.com/vieux/docker-volume-sshfs/tree/stateful

here is my usecase:

```
#install plugin
$ docker plugin install vieux/sshfs:next DEBUG=1

#create volume
$ docker volume create -d vieux/sshfs:next -o sshcmd=root@1.2.3.4:/tmp/shared -o password=XXX sshvolume

#use volume
$ docker run -it -v sshvolume:/data alpine sh -c "touch /data/hello"

#disable plugin but don't remove the volume
$ docker plugin disable -f vieux/sshfs:next

##here docker volume ls doesn't show 'sshfsvolume', since the plugin is disabled

#fake upgrade
$ docker plugin upgrade vieux/sshfs:next vieux/sshfs:next

#enable plugin
$ docker plugin enable vieux/sshfs:next

##here docker volume ls shows 'sshfsvolume'

#reuse the plugin 
$ docker run -it -v sshvolume:/data alpine sh -c "ls /data"
hello
```

I tried with the branch when the tests were GREEN, I'm tempted to say LGTM
I'll retry once the tests are green again.

cpuguy83 commented:
Fixed the issue by adding a recursive unmount on plugin stop and before plugin remove.

tiborvass commented on api/server/router/plugin/plugin_routes.go:
```diff
@@ -116,20 +116,77 @@ func (pr *pluginRouter) pullPlugin(ctx context.Context, w http.ResponseWriter, r
+	if err != nil {
+		return err
+	}
+	w.Header().Set("Docker-Plugin-Name", name)
```
Shouldn't this be `X-Docker-Plugin-Name` ?

vieux replied:
what's the point of this header ?

cpuguy83 replied:
@tiborvass I just copied this from the existing code: https://github.com/docker/docker/pull/29414/files#diff-3869b4c752615bcebd087e8cf5eee874L154

cpuguy83 replied:
Maybe I don't need it here

anusha-ragunathan commented:
This needs a `plugin_upgrade.md` https://github.com/docker/docker/tree/master/docs/reference/commandline with a working example. You can use https://github.com/docker/docker/pull/29414#issuecomment-277192456 for content.

cpuguy83 commented:
Added docs.

anusha-ragunathan commented:
This needs to be automated at some point, but for now, we need to add `.Image` to the list in https://github.com/docker/docker/blob/master/docs/reference/commandline/plugin_ls.md

anusha-ragunathan commented:
@vieux : Not something urgent, but we shouldnt be writing plugin specific state files in the directory reserved for docker internals. i.e dont create `sshfs-state.json` in `/var/lib/docker/plugins`

anusha-ragunathan commented on docs/reference/commandline/plugin_upgrade.md:
````diff
@@ -0,0 +1,85 @@
+```markdown
+Usage:  docker plugin upgrade [OPTIONS] PLUGIN [REMOTE]
+
+Install a plugin
````
Shod be "Upgrade an existing plugin"

anusha-ragunathan commented on docs/reference/commandline/plugin_upgrade.md:
```diff
@@ -0,0 +1,85 @@
+Install a plugin
+
+Options:
+      --alias string            Local name for plugin
```
there's no alias option for upgrade

anusha-ragunathan commented:
Might be worth documenting the "bind mount on host outside rootfs" behavior of `PropagateMount` in https://github.com/docker/docker/blob/master/docs/extend/config.md.

vieux commented:
@cpuguy83 test failure

vieux commented:
LGTM we might add swagger and fix doc after rc2

cpuguy83 commented:
Oh darn I missed the new endpoint for swagger... yeah I can take care of this after.

tiborvass commented:
TestStackRemove is flakey, disregard.

anusha-ragunathan commented:
LGTM

thaJeztah commented:
cherry-picked into 1.13.x through https://github.com/docker/docker/pull/30728
